File: src/core/dbc-parser.js
// Vector DBC text parser.
//
// Supported statements: VERSION, BU_, BO_, SG_ (incl. simple multiplexing
// m<N>/M), VAL_, VAL_TABLE_, CM_ (multi-line, for BU_/BO_/SG_/global),
// BA_DEF_DEF_ / BA_ (GenSigStartValue -> Signal.startValue), SIG_VALTYPE_
// (float/double signals). Extended multiplexing (SG_MUL_VAL_) is recorded as a
// warning, not decoded. Anything unrecognized is skipped silently; malformed
// versions of *recognized* statements are collected in cluster.parseErrors
// with their line number, and parsing continues.

import { Cluster, Message, Signal } from './dbc-model.js';

const DBC_EXT_FLAG = 0x80000000; // DBC stores extended IDs with bit 31 set

const RE = {
  bo: /^BO_\s+(\d+)\s+([A-Za-z_][\w-]*)\s*:\s*(\d+)\s+(\S+)\s*$/,
  sg: /^SG_\s+([A-Za-z_][\w-]*)\s*(m\d+M?|M)?\s*:\s*(\d+)\|(\d+)@([01])([+-])\s*\(\s*([^,\s]+)\s*,\s*([^)\s]+)\s*\)\s*\[\s*([^|]*?)\s*\|\s*([^\]]*?)\s*\]\s*"([^"]*)"\s*(.*)$/,
  val: /^VAL_\s+(\d+)\s+([A-Za-z_][\w-]*)\s+(.*?);?\s*$/,
  valTable: /^VAL_TABLE_\s+([A-Za-z_][\w-]*)\s+(.*?);?\s*$/,
  valPairs: /(-?\d+)\s+"((?:[^"\\]|\\.)*)"/g,
  cmSg: /^CM_\s+SG_\s+(\d+)\s+([A-Za-z_][\w-]*)\s+"([\s\S]*)"\s*;?\s*$/,
  cmBo: /^CM_\s+BO_\s+(\d+)\s+"([\s\S]*)"\s*;?\s*$/,
  cmBu: /^CM_\s+BU_\s+([A-Za-z_][\w-]*)\s+"([\s\S]*)"\s*;?\s*$/,
  cmGlobal: /^CM_\s+"([\s\S]*)"\s*;?\s*$/,
  baDefDef: /^BA_DEF_DEF_\s+"([^"]+)"\s+(.+?)\s*;?\s*$/,
  baSg: /^BA_\s+"([^"]+)"\s+SG_\s+(\d+)\s+([A-Za-z_][\w-]*)\s+(.+?)\s*;?\s*$/,
  baBo: /^BA_\s+"([^"]+)"\s+BO_\s+(\d+)\s+(.+?)\s*;?\s*$/,
  sigValtype: /^SIG_VALTYPE_\s+(\d+)\s+([A-Za-z_][\w-]*)\s*:?\s*(\d+)\s*;?\s*$/,
  bu: /^BU_\s*:\s*(.*)$/,
  sgMulVal: /^SG_MUL_VAL_\s/,
};

/**
 * Parse DBC text into a Cluster.
 * @param {string} text DBC file contents
 * @param {string} fileName original file name (cluster is named after it)
 */
export function parseDbc(text, fileName) {
  const name = fileName.replace(/\.[^.]*$/, '');
  const cluster = new Cluster(name, fileName);
  const byDbcId = new Map(); // raw DBC id (incl. ext flag) -> Message
  const valueTables = new Map(); // VAL_TABLE_ name -> Map
  const sigStartValues = []; // applied at the end (BA_ can precede nothing)
  let currentMessage = null;
  let extMuxWarned = false;

  const err = (line, textMsg) => cluster.parseErrors.push({ line, text: textMsg });

  const statements = splitStatements(text);
  for (const { line, stmt } of statements) {
    try {
      if (stmt.startsWith('BO_ ')) {
        const m = RE.bo.exec(stmt);
        if (!m) {
          err(line, `Malformed BO_ statement: ${truncate(stmt)}`);
          currentMessage = null;
          continue;
        }
        const dbcId = Number(m[1]) >>> 0;
        const extended = (dbcId & DBC_EXT_FLAG) !== 0;
        const msg = new Message({
          id: dbcId & ~DBC_EXT_FLAG,
          extended,
          name: m[2],
          dlc: Number(m[3]),
          transmitter: m[4] === 'Vector__XXX' ? '' : m[4],
        });
        byDbcId.set(dbcId, msg);
        cluster.messages.set(dbcId >>> 0, msg);
        currentMessage = msg;
      } else if (stmt.startsWith('SG_ ')) {
        if (!currentMessage) {
          err(line, 'SG_ outside of a BO_ block');
          continue;
        }
        const m = RE.sg.exec(stmt);
        if (!m) {
          err(line, `Malformed SG_ statement: ${truncate(stmt)}`);
          continue;
        }
        const mux = m[2] ?? '';
        const signal = new Signal({
          name: m[1],
          startBit: Number(m[3]),
          bitLength: Number(m[4]),
          byteOrder: Number(m[5]), // 1 = Intel, 0 = Motorola
          signed: m[6] === '-',
          factor: Number(m[7]),
          offset: Number(m[8]),
          min: m[9] === '' ? 0 : Number(m[9]),
          max: m[10] === '' ? 0 : Number(m[10]),
          unit: m[11],
          receivers: m[12]
            .split(/[,\s]+/)
            .filter((r) => r && r !== 'Vector__XXX'),
          muxRole: mux === 'M' ? 'multiplexor' : mux ? 'multiplexed' : 'none',
          muxValue: /^m\d+/.test(mux) ? Number(mux.match(/^m(\d+)/)[1]) : null,
        });
        currentMessage.signals.push(signal);
      } else if (stmt.startsWith('VAL_TABLE_ ')) {
        const m = RE.valTable.exec(stmt);
        if (m) valueTables.set(m[1], parseValuePairs(m[2]));
        else err(line, `Malformed VAL_TABLE_: ${truncate(stmt)}`);
      } else if (stmt.startsWith('VAL_ ')) {
        const m = RE.val.exec(stmt);
        if (!m) {
          err(line, `Malformed VAL_: ${truncate(stmt)}`);
          continue;
        }
        const sig = findSignal(byDbcId, Number(m[1]), m[2]);
        if (sig) {
          // Either inline pairs or a VAL_TABLE_ reference.
          const rest = m[3].trim();
          const table = valueTables.get(rest);
          sig.valueTable = table ?? parseValuePairs(rest);
          if (sig.valueTable.size === 0) sig.valueTable = null;
        }
      } else if (stmt.startsWith('CM_')) {
        let m;
        if ((m = RE.cmSg.exec(stmt))) {
          const sig = findSignal(byDbcId, Number(m[1]), m[2]);
          if (sig) sig.comment = unescapeQuotes(m[3]);
        } else if ((m = RE.cmBo.exec(stmt))) {
          const msg = byDbcId.get(Number(m[1]) >>> 0);
          if (msg) msg.comment = unescapeQuotes(m[2]);
        } else if ((m = RE.cmBu.exec(stmt))) {
          // node comments: not surfaced yet
        } else if ((m = RE.cmGlobal.exec(stmt))) {
          cluster.comment = unescapeQuotes(m[1]);
        }
      } else if (stmt.startsWith('BA_DEF_DEF_ ')) {
        const m = RE.baDefDef.exec(stmt);
        if (m && m[1] === 'GenSigStartValue') {
          const def = Number(stripQuotes(m[2]));
          if (!Number.isNaN(def) && def !== 0) {
            sigStartValues.push({ default: def });
          }
        }
      } else if (stmt.startsWith('BA_ ')) {
        let m;
        if ((m = RE.baSg.exec(stmt))) {
          if (m[1] === 'GenSigStartValue') {
            const sig = findSignal(byDbcId, Number(m[2]), m[3]);
            const v = Number(stripQuotes(m[4]));
            if (sig && !Number.isNaN(v)) sig.startValue = v;
          }
        } else if ((m = RE.baBo.exec(stmt))) {
          // message attributes (GenMsgCycleTime etc.) — not surfaced yet
        }
      } else if (stmt.startsWith('SIG_VALTYPE_ ')) {
        const m = RE.sigValtype.exec(stmt);
        if (m) {
          const sig = findSignal(byDbcId, Number(m[1]), m[2]);
          if (sig) {
            sig.isFloat = m[3] === '1' || m[3] === '2';
            sig.isDouble = m[3] === '2';
          }
        }
      } else if (RE.sgMulVal.test(stmt)) {
        if (!extMuxWarned) {
          err(line, 'Extended multiplexing (SG_MUL_VAL_) is not decoded; affected signals use simple m<N> selectors only.');
          extMuxWarned = true;
        }
      } else if (stmt.startsWith('BU_')) {
        const m = RE.bu.exec(stmt);
        if (m) {
          cluster.nodes = m[1].split(/\s+/).filter((n) => n && n !== 'Vector__XXX');
        }
      }
      // VERSION, NS_, BS_, BA_DEF_, EV_, SIG_GROUP_, ... intentionally ignored.
    } catch (e) {
      err(line, `Parser error: ${e.message}`);
    }
  }

  // Apply GenSigStartValue default to signals without an explicit BA_ value.
  const defEntry = sigStartValues.find((s) => 'default' in s);
  if (defEntry) {
    for (const msg of cluster.messages.values()) {
      for (const sig of msg.signals) {
        if (sig.startValue == null) sig.startValue = defEntry.default;
      }
    }
  }

  // Deterministic signal order: by start bit within each message.
  for (const msg of cluster.messages.values()) {
    msg.signals.sort((a, b) => a.startBit - b.startBit || a.name.localeCompare(b.name));
  }
  return cluster;
}

/**
 * Split DBC text into statements with their starting line numbers.
 * Statements are line-based, except quoted strings (comments!) may span
 * lines — so lines are joined while inside an open quote. SG_ lines belong to
 * the preceding BO_; the parser keeps that state, so here every physical
 * statement is emitted separately.
 */
function splitStatements(text) {
  const out = [];
  const lines = text.split(/\r\n|\r|\n/);
  let buf = '';
  let bufLine = 0;
  let inQuote = false;
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    if (!inQuote) {
      buf = line.trim();
      bufLine = i + 1;
      if (!buf) continue;
    } else {
      buf += '\n' + line;
    }
    inQuote = countUnescapedQuotes(buf) % 2 === 1;
    if (!inQuote) out.push({ line: bufLine, stmt: buf });
  }
  if (inQuote && buf.trim()) out.push({ line: bufLine, stmt: buf });
  return out;
}

function countUnescapedQuotes(s) {
  let n = 0;
  for (let i = 0; i < s.length; i++) {
    if (s[i] === '"' && s[i - 1] !== '\\') n++;
  }
  return n;
}

function parseValuePairs(s) {
  const table = new Map();
  RE.valPairs.lastIndex = 0;
  let m;
  while ((m = RE.valPairs.exec(s))) table.set(Number(m[1]), unescapeQuotes(m[2]));
  return table;
}

function findSignal(byDbcId, dbcId, sigName) {
  const msg = byDbcId.get(dbcId >>> 0);
  return msg?.signals.find((s) => s.name === sigName) ?? null;
}

function stripQuotes(s) {
  return s.replace(/^"(.*)"$/s, '$1');
}

function unescapeQuotes(s) {
  return s.replace(/\\"/g, '"');
}

function truncate(s, n = 80) {
  return s.length > n ? s.slice(0, n) + '…' : s;
}

File: python/canviewer/core/dbc_parser.py
"""Vector DBC text parser. Port of the JS dbc-parser.

Supports VERSION, BU_, BO_, SG_ (incl. simple m<N>/M multiplexing), VAL_,
VAL_TABLE_, CM_ (multi-line), BA_DEF_DEF_/BA_ GenSigStartValue, SIG_VALTYPE_.
Extended multiplexing (SG_MUL_VAL_) is recorded as a warning, not decoded.
"""
from __future__ import annotations

import re

from .dbc_model import Cluster, Message, Signal

DBC_EXT_FLAG = 0x80000000

RE_BO = re.compile(r"^BO_\s+(\d+)\s+([A-Za-z_][\w-]*)\s*:\s*(\d+)\s+(\S+)\s*$")
RE_SG = re.compile(
    r"^SG_\s+([A-Za-z_][\w-]*)\s*(m\d+M?|M)?\s*:\s*(\d+)\|(\d+)@([01])([+-])"
    r"\s*\(\s*([^,\s]+)\s*,\s*([^)\s]+)\s*\)\s*\[\s*([^|]*?)\s*\|\s*([^\]]*?)\s*\]"
    r'\s*"([^"]*)"\s*(.*)$'
)
RE_VAL = re.compile(r"^VAL_\s+(\d+)\s+([A-Za-z_][\w-]*)\s+(.*?);?\s*$")
RE_VALTABLE = re.compile(r"^VAL_TABLE_\s+([A-Za-z_][\w-]*)\s+(.*?);?\s*$")
RE_VALPAIRS = re.compile(r'(-?\d+)\s+"((?:[^"\\]|\\.)*)"')
RE_CM_SG = re.compile(r'^CM_\s+SG_\s+(\d+)\s+([A-Za-z_][\w-]*)\s+"([\s\S]*)"\s*;?\s*$')
RE_CM_BO = re.compile(r'^CM_\s+BO_\s+(\d+)\s+"([\s\S]*)"\s*;?\s*$')
RE_CM_GLOBAL = re.compile(r'^CM_\s+"([\s\S]*)"\s*;?\s*$')
RE_BADEFDEF = re.compile(r'^BA_DEF_DEF_\s+"([^"]+)"\s+(.+?)\s*;?\s*$')
RE_BA_SG = re.compile(r'^BA_\s+"([^"]+)"\s+SG_\s+(\d+)\s+([A-Za-z_][\w-]*)\s+(.+?)\s*;?\s*$')
RE_SIGVALTYPE = re.compile(r"^SIG_VALTYPE_\s+(\d+)\s+([A-Za-z_][\w-]*)\s*:?\s*(\d+)\s*;?\s*$")
RE_BU = re.compile(r"^BU_\s*:\s*(.*)$")
RE_SGMULVAL = re.compile(r"^SG_MUL_VAL_\s")


def _unescape(s: str) -> str:
    return s.replace('\\"', '"')


def _strip_quotes(s: str) -> str:
    m = re.match(r'^"(.*)"$', s, re.S)
    return m.group(1) if m else s


def _truncate(s, n=80):
    return s[:n] + "…" if len(s) > n else s


def _count_unescaped_quotes(s: str) -> int:
    n = 0
    for i, ch in enumerate(s):
        if ch == '"' and (i == 0 or s[i - 1] != "\\"):
            n += 1
    return n


def _split_statements(text: str):
    out = []
    lines = re.split(r"\r\n|\r|\n", text)
    buf = ""
    buf_line = 0
    in_quote = False
    for i, line in enumerate(lines):
        if not in_quote:
            buf = line.strip()
            buf_line = i + 1
            if not buf:
                continue
        else:
            buf += "\n" + line
        in_quote = _count_unescaped_quotes(buf) % 2 == 1
        if not in_quote:
            out.append((buf_line, buf))
    if in_quote and buf.strip():
        out.append((buf_line, buf))
    return out


def _parse_value_pairs(s: str) -> dict:
    return {int(m.group(1)): _unescape(m.group(2)) for m in RE_VALPAIRS.finditer(s)}


def _find_signal(by_id, dbc_id, name):
    msg = by_id.get(dbc_id & 0xFFFFFFFF)
    if not msg:
        return None
    for s in msg.signals:
        if s.name == name:
            return s
    return None


def parse_dbc(text: str, file_name: str) -> Cluster:
    name = re.sub(r"\.[^.]*$", "", file_name)
    cluster = Cluster(name, file_name)
    by_id: dict[int, Message] = {}
    value_tables: dict[str, dict] = {}
    default_start = None
    current = None
    ext_mux_warned = False

    def err(line, msg):
        cluster.parse_errors.append({"line": line, "text": msg})

    for line, stmt in _split_statements(text):
        try:
            if stmt.startswith("BO_ "):
                m = RE_BO.match(stmt)
                if not m:
                    err(line, f"Malformed BO_ statement: {_truncate(stmt)}")
                    current = None
                    continue
                dbc_id = int(m.group(1)) & 0xFFFFFFFF
                extended = bool(dbc_id & DBC_EXT_FLAG)
                msg = Message(
                    id=dbc_id & ~DBC_EXT_FLAG & 0xFFFFFFFF,
                    extended=extended,
                    name=m.group(2),
                    dlc=int(m.group(3)),
                    transmitter="" if m.group(4) == "Vector__XXX" else m.group(4),
                )
                by_id[dbc_id] = msg
                cluster.messages[dbc_id] = msg
                current = msg
            elif stmt.startswith("SG_ "):
                if current is None:
                    err(line, "SG_ outside of a BO_ block")
                    continue
                m = RE_SG.match(stmt)
                if not m:
                    err(line, f"Malformed SG_ statement: {_truncate(stmt)}")
                    continue
                mux = m.group(2) or ""
                mux_value = int(re.match(r"^m(\d+)", mux).group(1)) if re.match(r"^m\d+", mux) else None
                sig = Signal(
                    name=m.group(1),
                    start_bit=int(m.group(3)),
                    bit_length=int(m.group(4)),
                    byte_order=int(m.group(5)),
                    signed=m.group(6) == "-",
                    factor=float(m.group(7)),
                    offset=float(m.group(8)),
                    min=0.0 if m.group(9) == "" else float(m.group(9)),
                    max=0.0 if m.group(10) == "" else float(m.group(10)),
                    unit=m.group(11),
                    receivers=[r for r in re.split(r"[,\s]+", m.group(12)) if r and r != "Vector__XXX"],
                    mux_role="multiplexor" if mux == "M" else ("multiplexed" if mux else "none"),
                    mux_value=mux_value,
                )
                current.signals.append(sig)
            elif stmt.startswith("VAL_TABLE_ "):
                m = RE_VALTABLE.match(stmt)
                if m:
                    value_tables[m.group(1)] = _parse_value_pairs(m.group(2))
                else:
                    err(line, f"Malformed VAL_TABLE_: {_truncate(stmt)}")
            elif stmt.startswith("VAL_ "):
                m = RE_VAL.match(stmt)
                if not m:
                    err(line, f"Malformed VAL_: {_truncate(stmt)}")
                    continue
                sig = _find_signal(by_id, int(m.group(1)), m.group(2))
                if sig:
                    rest = m.group(3).strip()
                    table = value_tables.get(rest) or _parse_value_pairs(rest)
                    sig.value_table = table if table else None
            elif stmt.startswith("CM_"):
                m = RE_CM_SG.match(stmt)
                if m:
                    sig = _find_signal(by_id, int(m.group(1)), m.group(2))
                    if sig:
                        sig.comment = _unescape(m.group(3))
                    continue
                m = RE_CM_BO.match(stmt)
                if m:
                    msg = by_id.get(int(m.group(1)) & 0xFFFFFFFF)
                    if msg:
                        msg.comment = _unescape(m.group(2))
                    continue
                m = RE_CM_GLOBAL.match(stmt)
                if m:
                    cluster.comment = _unescape(m.group(1))
            elif stmt.startswith("BA_DEF_DEF_ "):
                m = RE_BADEFDEF.match(stmt)
                if m and m.group(1) == "GenSigStartValue":
                    try:
                        d = float(_strip_quotes(m.group(2)))
                        if d != 0:
                            default_start = d
                    except ValueError:
                        pass
            elif stmt.startswith("BA_ "):
                m = RE_BA_SG.match(stmt)
                if m and m.group(1) == "GenSigStartValue":
                    sig = _find_signal(by_id, int(m.group(2)), m.group(3))
                    try:
                        v = float(_strip_quotes(m.group(4)))
                        if sig:
                            sig.start_value = v
                    except ValueError:
                        pass
            elif stmt.startswith("SIG_VALTYPE_ "):
                m = RE_SIGVALTYPE.match(stmt)
                if m:
                    sig = _find_signal(by_id, int(m.group(1)), m.group(2))
                    if sig:
                        sig.is_float = m.group(3) in ("1", "2")
                        sig.is_double = m.group(3) == "2"
            elif RE_SGMULVAL.match(stmt):
                if not ext_mux_warned:
                    err(line, "Extended multiplexing (SG_MUL_VAL_) is not decoded.")
                    ext_mux_warned = True
            elif stmt.startswith("BU_"):
                m = RE_BU.match(stmt)
                if m:
                    cluster.nodes = [n for n in re.split(r"\s+", m.group(1)) if n and n != "Vector__XXX"]
        except Exception as e:  # noqa: BLE001 - collect and continue
            err(line, f"Parser error: {e}")

    if default_start is not None:
        for msg in cluster.messages.values():
            for sig in msg.signals:
                if sig.start_value is None:
                    sig.start_value = default_start

    for msg in cluster.messages.values():
        msg.signals.sort(key=lambda s: (s.start_bit, s.name))
    return cluster
